workflow/executor/resource.go

```diff
@@ -80,51 +92,151 @@ func (we *WorkflowExecutor) WaitResource(resourceName string) error {
 		failReqs, _ = failSelector.Requirements()
 	}
 
-	cmd := exec.Command("kubectl", "get", resourceName, "-w", "-o", "json")
-	stdout, err := cmd.StdoutPipe()
+	ctrlChannel := make(chan string)
+	resultChannel := make(chan waitResult)
+	go checkResourceState(ctrlChannel, resultChannel, resourceName, successReqs, failReqs)
+
+	err := wait.ExponentialBackoff(wait.Backoff{Duration: (time.Second * 5), Factor: 4.0, Steps: 5},
+		func() (bool, error) {
+			ctrlChannel <- startStr
+
+			result := <-resultChannel
+
+			if result.result == "" {
+				msg := fmt.Sprintf("Waiting for resource %s resulted in channel close", resourceName)
+				log.Warnf(msg)
+				return false, errors.InternalError(msg)
+			}
+
+			if result.err == nil {
+				log.Infof("Returning from successfull wait for resource %s", resourceName)
+				return true, nil
+			}
+
+			if result.result == retryStr {
+				log.Infof("Waiting for resource %s resulted in retryable error %v", resourceName, result.err)
+				return false, nil
+			}
+
+			log.Warnf("Waiting for resource %s resulted in non-retryable error %v", resourceName, result.err)
+			return false, result.err
+		})
+
+	close(ctrlChannel)
```

Comment:
This forces the go routine to stop executing and exit. Since after ExponentialBackoff returns we are done with all the iterations that we want to run. So when eventually the go routine goes back to waiting for the ctrlChannel it will see the close channel and exit itself.